File: valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> pair = Map.of('(', ')', '{', '}', '[', ']');
        Deque<Character> stack = new LinkedList<>();
           
        for(int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if (pair.containsKey(c)) {
                stack.push(pair.get(c));
            } else {
                if (stack.isEmpty() || stack.peek() != c) return false;
                stack.pop();
            }
        }
        
        return stack.isEmpty();
    }
}

File: valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        m = {
            '}': '{',
            ']': '[',
            ')': '('
        }
        
        stack = []
        
        for c in s:
            if c in m:
                if len(stack) == 0 or stack.pop() != m.get(c):
                    return False
                continue
            stack.append(c)
            
        return len(stack) == 0
        
        